docs/sample-apps/multimodal-chat/backend/api/routes.py

```diff
@@ -0,0 +1,583 @@
+    }]
+
+    # Add historical messages
+    messages.extend([{
+        'role': msg['role'],
+        'content': msg['content']
+    } for msg in history])
+
+    # Add current prompt
+    messages.append({
+        'role': 'user',
+        'content': prompt
+    })
+
+    return messages
+
+# Setup similarity search query
+@chunks_view.query
+def get_relevant_chunks(query_text: str):
+    sim = chunks_view.text.similarity(query_text)
+    return (
+        chunks_view.order_by(sim, asc=False).select(chunks_view.text, sim=sim).limit(20)
+    )
+
+
+@transcription_chunks.query
+def get_relevant_transcript_chunks(query_text: str):
+    sim = transcription_chunks.text.similarity(query_text)
+    return (
+        transcription_chunks.order_by(sim, asc=False)
+        .select(transcription_chunks.text, sim=sim)
+        .limit(20)
+    )
+
+
+@audio_chunks.query
+def get_relevant_audio_chunks(query_text: str):
+    sim = audio_chunks.text.similarity(query_text)
+    return (
+        audio_chunks.order_by(sim, asc=False)
+        .select(audio_chunks.text, sim=sim)
+        .limit(20)
+    )
+
+
+# Add computed columns
+docs_table["context_doc"] = chunks_view.queries.get_relevant_chunks(docs_table.question)
+docs_table["context_video"] = (
+    transcription_chunks.queries.get_relevant_transcript_chunks(docs_table.question)
+)
+docs_table["context_audio"] = audio_chunks.queries.get_relevant_audio_chunks(
+    docs_table.question
+)
+docs_table["prompt"] = create_prompt(
+    docs_table.context_doc,
+    docs_table.context_video,
+    docs_table.context_audio,
+    docs_table.question,
+)
+docs_table['chat_history'] = conversations.queries.get_chat_history()
+docs_table['messages'] = create_messages(
+    docs_table.chat_history,
+    docs_table.prompt
+)
+docs_table["response"] = openai.chat_completions(
+    messages=docs_table.messages,
+    model="gpt-4o-mini",
+)
+
+# Extract the answer text from the API response
+docs_table["answer"] = docs_table.response.choices[0].message.content
+
+logger.info("Setup complete")
+
+
+async def process_document(file_path: Path) -> bool:
+    try:
+        if file_path.suffix.lower() == ".py":
+            # Create markdown file with Python syntax highlighting
+            md_path = file_path.with_suffix(".md")
+            with open(file_path, "r", encoding="utf-8") as py_file:
+                content = py_file.read()
+            with open(md_path, "w", encoding="utf-8") as md_file:
+                md_file.write(f"```python\n{content}\n```")
+            file_path = md_path
+
+        abs_path = str(file_path.absolute()).replace(os.sep, "/")
+        logger.info(f"Processing document: {abs_path}")
+
+        docs_table = pxt.get_table("chatbot.documents")
+        docs_table.insert([{"document": abs_path}])
+        return True
+
+    except Exception as e:
+        logger.error(f"Error processing document: {e}")
+        raise
+
+
+@router.post("/api/upload")
```

Comment:
let me come back to you on that one as we should (1) remove the check for video file that is superfluous now that we have the other api/videos/upload and want to see if I can get the .py -> .md working if not we can also ditch that code.